shared/rollouts/__init__.py

```diff
@@ -106,45 +114,83 @@ def _buckets(self):
         this class will compute a hash including:
         - The experiment's name
         - The repo's id
-        - (Optional) A unique salt
+        - A salt
 
         The range of possible hash values is divvied up into buckets based on the
         `proportion` and `variants` members. The hash for this repo will fall into
         one of those buckets and the corresponding variant (or default value) will
         be returned.
         """
-        test_population = int(self.proportion * Feature.HASHSPACE)
+        test_population = int(self.feature_flag.proportion * Feature.HASHSPACE)
 
         buckets = []
         quantile = 0
-        for variant in self.variants.values():
+        for variant in self.ff_variants:
             quantile += variant.proportion
             buckets.append((int(quantile * test_population), variant))
 
         return buckets
 
-    # NOTE: `@lru_cache` on instance methods is ordinarily a bad idea:
+    def _get_override_variant(self, identifier):
+        """
+        Retrieves the FeatureFlagVariant applicable to the given identifer according to
+        defined overrides. Returns None if no override is found.
+        """
+        for variant in self.ff_variants:
+            if (
+                identifier in variant.override_owner_ids
+                or identifier in variant.override_repo_ids
+            ):
+                return variant
+        return None
+
+    def _is_valid_rollout(self):
+        """
+        Checks if the FeatureFlagVariant entries were given invalid values, which is very
+        possible since these values can be modified via Django Admin. Due to the TTL cache,
+        the values within this class will only be updated once every 5 minutes.
+        """
+        return sum(map(lambda x: x.proportion, self.ff_variants)) == 1.0 and not (
+            True
+            in map(lambda x: x.proportion < 0 or x.proportion > 1, self.ff_variants)
+        )
+
+    # NOTE: `@ttl_cache` on instance methods is ordinarily a bad idea:
     # - the cache is not per-instance; it's shared by all class instances
     # - by holding references to function arguments in the cache, it prolongs
     #   their lifetimes. Since `self` is a function argument, the cache will
     #   prevent any instance from getting freed until it is evicted from the
     #   cache which could be a significant memory leak.
     # In this case, we are okay with sharing a cache across instances, and the
     # instances are all global constants so they won't be torn down anyway.
-    @lru_cache(maxsize=50)
-    def check_value(self, identifier, default=None):
-        identifer = str(identifier)  # just in case
-        if identifier in self.overrides:
-            override_variant = self.overrides[identifier]
-            return self.variants[override_variant].value
-        elif self.proportion == 1.0 and len(self.variants) == 1:
+    @ttl_cache(maxsize=64, ttl=300)  # 5 minute ttl
```

Comment:
hmm. i think this will wind up making more database queries than we want. consider the following:
- we call `feature.check_value(A)`. we'll fetch the model from the db, compute which variant `A` belongs in, put it in the cache, and then return
- we immediately call `feature.check_value(B)`. `B` isn't in our cache, so we run the function again. we make a db request, compute the variant, cache it, and then return

what i think we want instead is to wrap accesses to the model in a `get_model()` function or something and put that behind a second cache, and the second cache has a TTL. something like:
```
# this caches the variant that each identifier should use, assuming the db
# model is up to date
@lru_cache(maxsize=64)
def check_value(self, identifier, default=None):
    model = self.model
    override_variant = self.get_override_variant(model)
    if override_variant:
        ... # and so on

# this will re-fetch the db model if it's stale. if anything in the model changed,
# we should clear the check_value() cache
#
# the @property means it's accessed as if it's a member variable: self.model
@property
@ttl_cache(maxsize=64, ttl=300)
def model(self):
    new_model = self.fetch_from_db()
    if new_model != self._model:
        self._model = new_model
        self.check_value.cache_clear()
    return self._model
```

Reply:
ah I see my mistake, thanks!